File: frontend/src/lib/urlstate.ts
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// TypeScript mirror of `almagest-query::urlstate` — encode the current parameter
// values into a shareable query string and decode them back, typed by the
// declarations. Keeping the wire format identical to the Rust side means a link
// shared from the viewer decodes the same whether the next open is served by
// Almagest or by the interactive-HTML export.
//
// Encoding is decl-aware so the query string stays human-readable
// (`region=EU`, not a JSON blob); a `multiselect` repeats its key, a
// `daterange` encodes its `preset` or its `start`/`end`. Parameters declared
// `persist: "none"` are omitted from the URL.

import type { Parameter } from "./types";

/** Build the `key=value&…` query string (no leading `?`) for the given values. */
export function encodeUrlState(
  values: Record<string, unknown>,
  decls: Parameter[],
): string {
  const parts: string[] = [];
  for (const decl of decls) {
    if (decl.persist === "none") continue;
    const value = values[decl.id];
    if (value === undefined || value === null || value === "") continue;

    if (Array.isArray(value)) {
      for (const item of value) {
        const s = scalarToString(item);
        if (s !== null) parts.push(pair(decl.id, s));
      }
    } else if (typeof value === "object") {
      const obj = value as Record<string, unknown>;
      const preset = typeof obj.preset === "string" ? obj.preset : null;
      if (preset) {
        parts.push(pair(`${decl.id}.preset`, preset));
      } else {
        if (typeof obj.start === "string" && obj.start) parts.push(pair(`${decl.id}.start`, obj.start));
        if (typeof obj.end === "string" && obj.end) parts.push(pair(`${decl.id}.end`, obj.end));
      }
    } else {
      const s = scalarToString(value);
      if (s !== null) parts.push(pair(decl.id, s));
    }
  }
  return parts.join("&");
}

/**
 * Decode a query string into raw parameter values, typed by `decls`. Keys not
 * matching any declaration are ignored; a value that doesn't parse for its kind
 * is dropped (so the caller falls back to the declared default).
 */
export function decodeUrlState(query: string, decls: Parameter[]): Record<string, unknown> {
  const raw = parseQuery(query);
  const out: Record<string, unknown> = {};

  for (const decl of decls) {
    const id = decl.id;
    switch (decl.kind) {
      case "multiselect": {
        const vals = raw.get(id);
        if (vals) out[id] = [...vals];
        break;
      }
      case "daterange": {
        const preset = raw.get(`${id}.preset`)?.[0];
        if (preset) {
          out[id] = { preset };
        } else {
          const start = raw.get(`${id}.start`)?.[0];
          const end = raw.get(`${id}.end`)?.[0];
          if (start && end) out[id] = { start, end };
        }
        break;
      }
      case "number": {
        const s = raw.get(id)?.[0];
        if (s !== undefined) {
          const n = Number(s);
          if (Number.isFinite(n)) out[id] = n;
        }
        break;
      }
      case "boolean": {
        const s = raw.get(id)?.[0];
        if (s === "true") out[id] = true;
        else if (s === "false") out[id] = false;
        break;
      }
      default: {
        // text | date | select
        const s = raw.get(id)?.[0];
        if (s !== undefined) out[id] = s;
        break;
      }
    }
  }
  return out;
}

/** Layer parameter sources: URL state overrides seeded defaults. */
export function layeredState(
  url: Record<string, unknown>,
  defaults: Record<string, unknown>,
): Record<string, unknown> {
  return { ...defaults, ...url };
}

function pair(key: string, value: string): string {
  return `${pctEncode(key)}=${pctEncode(value)}`;
}

function scalarToString(v: unknown): string | null {
  if (typeof v === "string") return v;
  if (typeof v === "number") return String(v);
  if (typeof v === "boolean") return String(v);
  return null;
}

/** Parse `key=value&…` into key → ordered values (repeated keys accumulate). */
function parseQuery(query: string): Map<string, string[]> {
  const out = new Map<string, string[]>();
  const q = query.startsWith("?") ? query.slice(1) : query;
  if (!q) return out;
  for (const pairStr of q.split("&")) {
    if (!pairStr) continue;
    const eq = pairStr.indexOf("=");
    const k = pctDecode(eq >= 0 ? pairStr.slice(0, eq) : pairStr);
    const v = eq >= 0 ? pctDecode(pairStr.slice(eq + 1)) : "";
    const arr = out.get(k);
    if (arr) arr.push(v);
    else out.set(k, [v]);
  }
  return out;
}

/** Percent-encode a query component, keeping the RFC 3986 unreserved set literal. */
function pctEncode(s: string): string {
  let out = "";
  for (const byte of new TextEncoder().encode(s)) {
    if (
      (byte >= 0x41 && byte <= 0x5a) || // A-Z
      (byte >= 0x61 && byte <= 0x7a) || // a-z
      (byte >= 0x30 && byte <= 0x39) || // 0-9
      byte === 0x2d || // -
      byte === 0x5f || // _
      byte === 0x2e || // .
      byte === 0x7e // ~
    ) {
      out += String.fromCharCode(byte);
    } else {
      out += "%" + byte.toString(16).toUpperCase().padStart(2, "0");
    }
  }
  return out;
}

/** Decode percent-escapes; `+` is treated as a space (form convention). */
function pctDecode(s: string): string {
  const bytes: number[] = [];
  for (let i = 0; i < s.length; ) {
    const c = s.charCodeAt(i);
    if (c === 0x25 /* % */ && i + 2 < s.length) {
      const h = parseInt(s.slice(i + 1, i + 3), 16);
      if (Number.isFinite(h)) {
        bytes.push(h);
        i += 3;
        continue;
      }
      bytes.push(c);
      i += 1;
    } else if (c === 0x2b /* + */) {
      bytes.push(0x20);
      i += 1;
    } else {
      bytes.push(c);
      i += 1;
    }
  }
  return new TextDecoder().decode(new Uint8Array(bytes));
}

File: crates/almagest-query/src/urlstate.rs
// SPDX-License-Identifier: MIT OR Apache-2.0

//! URL parameter state (doc 07): encode the current parameter values into a
//! shareable query string and decode them back, typed by the declarations.
//!
//! URL persistence is what makes a filtered view link-shareable ("here's the
//! dashboard with region=EU and the last 90 days"). Encoding is **decl-aware**
//! so the query string stays human-readable (`region=EU`, not a blob of JSON);
//! decoding consults the [`Parameter`] kinds to rebuild the right JSON shape
//! (a `multiselect` becomes an array, a `daterange` an object).
//!
//! A small percent-encoder is included rather than a dependency — the query
//! engine has no other use for a URL crate, and minimizing the dependency
//! surface is a project value.

use almagest_core::{ParamKind, Parameter};
use serde_json::{Map, Value};
use std::collections::BTreeMap;

/// Encode parameter values (keyed by id) into a `key=value&…` query string
/// (without a leading `?`). Scalars encode as themselves; a `multiselect`
/// array repeats its key; a `daterange` object encodes either its `preset` or
/// its `start`/`end` endpoints. `null` values are omitted.
pub fn encode_url_state(values: &BTreeMap<String, Value>) -> String {
    let mut parts: Vec<String> = Vec::new();
    for (id, value) in values {
        match value {
            Value::Null => {}
            Value::Array(items) => {
                for item in items {
                    if let Some(s) = scalar_to_string(item) {
                        parts.push(pair(id, &s));
                    }
                }
            }
            Value::Object(obj) => {
                if let Some(preset) = obj.get("preset").and_then(Value::as_str) {
                    parts.push(pair(&format!("{id}.preset"), preset));
                } else {
                    if let Some(s) = obj.get("start").and_then(Value::as_str) {
                        parts.push(pair(&format!("{id}.start"), s));
                    }
                    if let Some(e) = obj.get("end").and_then(Value::as_str) {
                        parts.push(pair(&format!("{id}.end"), e));
                    }
                }
            }
            other => {
                if let Some(s) = scalar_to_string(other) {
                    parts.push(pair(id, &s));
                }
            }
        }
    }
    parts.join("&")
}

/// Decode a query string into raw parameter values, typed by `decls`. Keys not
/// matching any declaration are ignored; a value that doesn't parse for its
/// kind is dropped (so resolution falls back to the declared default). The
/// result feeds [`crate::resolve::resolve_parameters`].
pub fn decode_url_state(query: &str, decls: &[Parameter]) -> BTreeMap<String, Value> {
    let raw = parse_query(query);
    let mut out = BTreeMap::new();

    for decl in decls {
        let id = &decl.id;
        match decl.kind {
            ParamKind::MultiSelect => {
                if let Some(values) = raw.get(id) {
                    let arr = values.iter().cloned().map(Value::String).collect();
                    out.insert(id.clone(), Value::Array(arr));
                }
            }
            ParamKind::DateRange => {
                if let Some(preset) = raw.get(&format!("{id}.preset")).and_then(|v| v.first()) {
                    let mut obj = Map::new();
                    obj.insert("preset".into(), Value::String(preset.clone()));
                    out.insert(id.clone(), Value::Object(obj));
                } else {
                    let start = raw.get(&format!("{id}.start")).and_then(|v| v.first());
                    let end = raw.get(&format!("{id}.end")).and_then(|v| v.first());
                    if let (Some(start), Some(end)) = (start, end) {
                        let mut obj = Map::new();
                        obj.insert("start".into(), Value::String(start.clone()));
                        obj.insert("end".into(), Value::String(end.clone()));
                        out.insert(id.clone(), Value::Object(obj));
                    }
                }
            }
            ParamKind::Number => {
                if let Some(s) = raw.get(id).and_then(|v| v.first())
                    && let Some(n) = parse_number(s)
                {
                    out.insert(id.clone(), n);
                }
            }
            ParamKind::Boolean => {
                if let Some(s) = raw.get(id).and_then(|v| v.first())
                    && let Ok(b) = s.parse::<bool>()
                {
                    out.insert(id.clone(), Value::Bool(b));
                }
            }
            ParamKind::Text | ParamKind::Date | ParamKind::Select => {
                if let Some(s) = raw.get(id).and_then(|v| v.first()) {
                    out.insert(id.clone(), Value::String(s.clone()));
                }
            }
        }
    }
    out
}

/// Layer parameter sources by priority: URL state overrides file-backed
/// defaults. (Declared defaults are applied later, by resolution, when a
/// parameter is absent from the merged map — completing the URL > file >
/// default precedence from doc 07.)
pub fn layered_state(
    url: &BTreeMap<String, Value>,
    file: &BTreeMap<String, Value>,
) -> BTreeMap<String, Value> {
    let mut out = file.clone();
    for (k, v) in url {
        out.insert(k.clone(), v.clone());
    }
    out
}

fn pair(key: &str, value: &str) -> String {
    format!("{}={}", pct_encode(key), pct_encode(value))
}

fn scalar_to_string(v: &Value) -> Option<String> {
    match v {
        Value::String(s) => Some(s.clone()),
        Value::Number(n) => Some(n.to_string()),
        Value::Bool(b) => Some(b.to_string()),
        _ => None,
    }
}

fn parse_number(s: &str) -> Option<Value> {
    if let Ok(i) = s.parse::<i64>() {
        Some(Value::Number(i.into()))
    } else if let Ok(f) = s.parse::<f64>() {
        serde_json::Number::from_f64(f).map(Value::Number)
    } else {
        None
    }
}

/// Parse a `key=value&…` query string into a map of key → ordered values
/// (repeated keys accumulate, supporting multiselect).
fn parse_query(query: &str) -> BTreeMap<String, Vec<String>> {
    let mut out: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let query = query.strip_prefix('?').unwrap_or(query);
    for pair in query.split('&') {
        if pair.is_empty() {
            continue;
        }
        let (k, v) = match pair.split_once('=') {
            Some((k, v)) => (pct_decode(k), pct_decode(v)),
            None => (pct_decode(pair), String::new()),
        };
        out.entry(k).or_default().push(v);
    }
    out
}

/// Percent-encode for a query component, keeping the RFC 3986 unreserved set
/// (`A–Z a–z 0–9 - _ . ~`) literal and escaping everything else.
fn pct_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        if b.is_ascii_alphanumeric() || matches!(b, b'-' | b'_' | b'.' | b'~') {
            out.push(b as char);
        } else {
            out.push('%');
            out.push(hex_digit(b >> 4));
            out.push(hex_digit(b & 0x0f));
        }
    }
    out
}

/// Decode percent-escapes; `+` is treated as a space (form convention).
fn pct_decode(s: &str) -> String {
    let bytes = s.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        match bytes[i] {
            b'%' if i + 2 < bytes.len() => match (hex_val(bytes[i + 1]), hex_val(bytes[i + 2])) {
                (Some(h), Some(l)) => {
                    out.push((h << 4) | l);
                    i += 3;
                }
                _ => {
                    out.push(b'%');
                    i += 1;
                }
            },
            b'+' => {
                out.push(b' ');
                i += 1;
            }
            b => {
                out.push(b);
                i += 1;
            }
        }
    }
    String::from_utf8_lossy(&out).into_owned()
}

fn hex_digit(n: u8) -> char {
    match n {
        0..=9 => (b'0' + n) as char,
        _ => (b'A' + (n - 10)) as char,
    }
}

fn hex_val(b: u8) -> Option<u8> {
    match b {
        b'0'..=b'9' => Some(b - b'0'),
        b'a'..=b'f' => Some(b - b'a' + 10),
        b'A'..=b'F' => Some(b - b'A' + 10),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use almagest_core::ParamKind;
    use serde_json::json;

    fn decl(id: &str, kind: ParamKind) -> Parameter {
        Parameter {
            id: id.into(),
            kind,
            label: None,
            description: None,
            default: None,
            options: None,
            options_query: None,
            min: None,
            max: None,
            min_selections: None,
            max_selections: None,
            allow_all: false,
            persist: None,
        }
    }

    #[test]
    fn percent_round_trip_handles_specials() {
        let s = "a b&c=d/e";
        assert_eq!(pct_decode(&pct_encode(s)), s);
    }

    #[test]
    fn decode_types_by_declaration() {
        let decls = vec![
            decl("region", ParamKind::Select),
            decl("limit", ParamKind::Number),
            decl("active", ParamKind::Boolean),
            decl("cats", ParamKind::MultiSelect),
            decl("dr", ParamKind::DateRange),
        ];
        let q = "region=EU&limit=50&active=true&cats=a&cats=b&dr.preset=last_30_days";
        let state = decode_url_state(q, &decls);
        assert_eq!(state["region"], json!("EU"));
        assert_eq!(state["limit"], json!(50));
        assert_eq!(state["active"], json!(true));
        assert_eq!(state["cats"], json!(["a", "b"]));
        assert_eq!(state["dr"], json!({ "preset": "last_30_days" }));
    }

    #[test]
    fn encode_decode_round_trip() {
        let decls = vec![
            decl("region", ParamKind::Select),
            decl("cats", ParamKind::MultiSelect),
            decl("dr", ParamKind::DateRange),
        ];
        let mut state = BTreeMap::new();
        state.insert("region".to_string(), json!("North & South"));
        state.insert("cats".to_string(), json!(["x", "y"]));
        state.insert(
            "dr".to_string(),
            json!({ "start": "2026-01-01", "end": "2026-03-31" }),
        );

        let encoded = encode_url_state(&state);
        let decoded = decode_url_state(&encoded, &decls);
        assert_eq!(decoded["region"], json!("North & South"));
        assert_eq!(decoded["cats"], json!(["x", "y"]));
        assert_eq!(
            decoded["dr"],
            json!({ "start": "2026-01-01", "end": "2026-03-31" })
        );
    }

    #[test]
    fn url_overrides_file_in_layering() {
        let mut file = BTreeMap::new();
        file.insert("region".to_string(), json!("US"));
        file.insert("limit".to_string(), json!(10));
        let mut url = BTreeMap::new();
        url.insert("region".to_string(), json!("EU"));

        let merged = layered_state(&url, &file);
        assert_eq!(merged["region"], json!("EU")); // URL wins
        assert_eq!(merged["limit"], json!(10)); // file retained
    }
}
